### ISSUE TITLE
Cannot use JobsClient in Management SDK to start a bulk user import job

### ISSUE THREAD

**Issue by sam-piper** [auth0/auth0.net#443]
### Description

I'm trying to use the latest version of the Management and Authentication SDKs (7.3.0) to perform a bulk user migration from a local SQL Server database into the standard Auth0 database connection (Username-Password-Authentication), using  `ManagementApiClient.Jobs.ImportUsersAsync()`.

When I call this method, I get back a `pending` job successfully, but the next time I check the job status, the job has failed, and there is no summary object or error message available in the response. 

If I call the job error details endpoint (`https://{domain}/api/v2/jobs/{job.Id}/errors`) manually, I get a 204 No Content.

If I look at the logs in the Auth0 Dashboard for the tenant I'm using, I can see the following (client_id / tenant redacted):

```
{
  "date": "2020-10-28T04:23:36.218Z",
  "type": "sapi",
  "description": "Create import users job",
  "client_id": "**********",
  "client_name": "",
  "ip": "220.245.220.52",
  "details": {
    "request": {
      "method": "post",
      "path": "/api/v2/jobs/users-imports",
      "query": {},
      "body": {
        "external_id": "UserImport-2020-10-28-15-23.json"
      },
      "channel": "api",
      "ip": "220.245.220.52",
      "auth": {
        "user": {},
        "strategy": "jwt",
        "credentials": {
          "scopes": [
            "read:users",
            "update:users",
            "delete:users",
            "create:users",
            "read:users_app_metadata",
            "update:users_app_metadata",
            "delete:users_app_metadata",
            "create:users_app_metadata"
          ]
        }
      }
    },
    "response": {
      "statusCode": 202,
      "body": {
        "id": "job_wxvyNkqAh5YLYsRu",
        "connection": "Username-Password-Authentication",
        "status": "pending",
        "type": "users_import",
        "external_id": "UserImport-2020-10-28-15-23.json",
        "created_at": "2020-10-28T04:23:36.201Z"
      }
    }
  },
  "auth0_client": {
    "name": "Auth0.Net",
    "version": "7.3.0",
    "env": {
      "target": "NETSTANDARD2.0"
    }
  },
  "log_id": "90020201028042337131000951970051294596893111385480757314",
  "_id": "90020201028042337131000951970051294596893111385480757314",
  "isMobile": false,
  "user_agent": "Other 0.0.0 / Other 0.0.0"
}
```

Which verifies that my request is being accepted and processed at the API layer, and there is also a log entry saying the job has completed:

```
{
  "date": "2020-10-28T04:23:36.372Z",
  "type": "sui",
  "description": "Users import finished",
  "connection": "Username-Password-Authentication",
  "connection_id": "",
  "details": {
    "tenant": "*******",
    "connection": "Username-Password-Authentication"
  },
  "log_id": "90020201028042341531000951973974258881542583400999682114",
  "_id": "90020201028042341531000951973974258881542583400999682114",
  "isMobile": false,
  "user_agent": "Other 0.0.0 / Other 0.0.0"
}
```

But there are no error details that help me, and the users are definitely not created.

The access token I'm obtaining has the required scopes as indicated in the API documentation - I've also used an access token that grants ALL Management API scopes, so it can't be an issue of missing scopes or an invalid token.

I've also installed the `Users Import / Export` extension, and when I drop my JSON file into this extension and run it, it completes successfully and the users get created with correct configuration (including password hashes), so it can't be the JSON file that is incorrect either.

### Reproduction

Here is a sample of the JSON file for a single user with modified email address, which runs successfully for me via the Users Import / Export extension:

```
[
  {
    "email": "test@user.com",
    "email_verified": true,
    "username": "Test.User",
    "blocked": false,
    "custom_password_hash": {
      "algorithm": "pbkdf2",
      "hash": {
        "value": "$pbkdf2-sha1$i=1000,l=32$d16owXF2G+jw/LpS122jDA$KgqXKR8tdXkTE31rrCsUjVkaDQiPkQQEGd4+u4e3V9g",
        "encoding": "utf8"
      }
    }
  }
]
```

Here is the C# code I'm using for this process - I couldn't find examples of this process in C# so had to work it out manually, so there may be problems here that I'm not aware of. Just drop into a new .NET Core 3.1 Console App, and add the required NuGets:

```
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Text;
using System.Threading.Tasks;
using Auth0.AuthenticationApi;
using Auth0.AuthenticationApi.Models;
using Auth0.ManagementApi;
using Auth0.ManagementApi.Models;
using Newtonsoft.Json;

namespace AisirisUserMigrator
{
    public class Program
    {
        private static readonly IManagementConnection _managementConnection = new HttpClientManagementConnection();
        private static readonly IAuthenticationConnection _authenticationConnection = new HttpClientAuthenticationConnection();
        private static string _domain;
        private static string _managementClientId;
        private static string _managementClientSecret;
        private static string _connectionId;
        private static string _accessToken;
        private static List<UserProfile> _users;
        private static ManagementApiClient _client;

        public static async Task<int> Main(string[] args)
        {
            try
            {
                // these need to be set appropriately
                _domain = "*****";
                _managementClientId = "*****";
                _managementClientSecret = "*****";
                _connectionId = "*****";

                _client = await GetManagementApiClient();

                // ISSUES:
                // The bulk import job is not currently working - but the same import file is working when uploaded via 
                // the Users Import / Export extension in Auth, and the access token appears to be valid.
                await MigrateUsers();
            }
            catch (Exception ex)
            {
                Console.WriteLine("An error occurred during execution:");
                Console.WriteLine(ex);
                Console.WriteLine("\r\nPress any key to exit...");
                Console.ReadKey();
                return 1;
            }

            return 0;
        }

        private static async Task MigrateUsers()
        {
            // this code can be used to verify a single profile's password hash migration
            _users = new List<UserProfile>
            {
                // Password for this hash should be bH9BOUvdmQKJ6oppm90nRWn4Y
                new UserProfile
                {
                    UserId = 1,
                    Email = "test@user.com",
                    UserName = "Test.User",
                    PasswordHash = "AHdeqMFxdhvo8Py6UtdtowwqCpcpHy11eRMTfWusKxSNWRoNCI+RBAQZ3j67h7dX2A=="
                }
            };

            var usersToImport = _users.Select(GetUserData).ToArray();

            // load users to import into JSON file
            var jsonFile = $"UserImport-{DateTime.Now:yyyy-MM-dd-HH-mm}.json";
            var jsonContent = JsonConvert.SerializeObject(usersToImport);
            await File.AppendAllTextAsync(jsonFile, jsonContent, Encoding.UTF8);
            Console.WriteLine("Created JSON import file '{0}' for {1} user records.", jsonFile, usersToImport.Length);

            // start a new import job using the JSON file
            Job job;
            await using (var jsonStream = File.OpenRead(jsonFile))
            {
                job = await _client.Jobs.ImportUsersAsync(_connectionId, jsonFile, jsonStream, true, jsonFile, false);
            }

            // wait for the job to finish on 2-second intervals
            do
            {
                await Task.Delay(2000);
                job = await _client.Jobs.GetAsync(job.Id);

                Console.WriteLine("Job status: {0}, Percentage complete: {1}%", job.Status, job.PercentageDone);
            }
            while (job.Status == "pending" || job.Status == "processing");

            // if the job has failed then emit summary details
            if (job.Status == "failed")
            {
                Console.WriteLine("Import job '{0}' did not complete successfully.", jsonFile);
                if (job.Summary != null)
                {
                    Console.WriteLine("Inserted: {0}, Updated: {1}, Failed: {2}, Total: {3}",
                                      job.Summary.Inserted, job.Summary.Updated, job.Summary.Failed, job.Summary.Total);
                }

                await LogJobErrorDetails(job);
            }
        }

        private static dynamic GetUserData(UserProfile user)
        {
            var (salt, hash) = ExtractSaltAndHash(user.PasswordHash);
            return new
            {
                email = user.Email,
                email_verified = true,
                username = user.UserName,
                blocked = false,
                custom_password_hash = new
                {
                    algorithm = "pbkdf2",
                    hash = new
                    {
                        value = $"$pbkdf2-sha1$i=1000,l=32${salt}${hash}",
                        encoding = "utf8"
                    }
                }
            };
        }

        private static (string, string) ExtractSaltAndHash(string passwordHash)
        {
            if (string.IsNullOrWhiteSpace(passwordHash))
            {
                throw new ArgumentNullException(nameof(passwordHash));
            }

            var numArray = Convert.FromBase64String(passwordHash);
            if (numArray.Length != 49 || numArray[0] != 0)
            {
                throw new ArgumentException("Password Hash is not a valid Base64 string.");
            }

            var salt = new byte[16];
            Buffer.BlockCopy(numArray, 1, salt, 0, 16);
            var encodedSalt = Convert.ToBase64String(salt).Replace("=", "");

            var hash = new byte[32];
            Buffer.BlockCopy(numArray, 17, hash, 0, 32);
            var encodedHash = Convert.ToBase64String(hash).Replace("=", "");

            return (encodedSalt, encodedHash);
        }

        private static async Task LogJobErrorDetails(Job job)
        {
            // there is no SDK method for retrieving error details (sigh) so have to use HttpClient directly
            using var httpClient = new HttpClient();
            httpClient.DefaultRequestHeaders.Authorization = new AuthenticationHeaderValue("Bearer", _accessToken);

            var response = await httpClient.GetAsync(new Uri($"https://{_domain}/api/v2/jobs/{job.Id}/errors"));
            response.EnsureSuccessStatusCode();

            var content = await response.Content.ReadAsStringAsync();
            var results = (dynamic[])JsonConvert.DeserializeObject(content);

            if (results?.Length > 0)
            {
                foreach (var result in results)
                {
                    Console.WriteLine("User with email '{0}' had the following errors:", result.user.email);
                    foreach (var error in result.errors)
                    {
                        Console.WriteLine("- {0}", error);
                    }

                    Console.WriteLine();
                }
            }
        }

        private static async Task<ManagementApiClient> GetManagementApiClient()
        {
            _accessToken = await GetAccessTokenForManagementApi();
            var client = new ManagementApiClient(_accessToken, _domain, _managementConnection);
            return client;
        }

        private static async Task<string> GetAccessTokenForManagementApi()
        {
            using var client = new AuthenticationApiClient(new Uri($"https://{_domain}"), _authenticationConnection);

            var tokenRequest = new ClientCredentialsTokenRequest
            {
                Audience = $"https://{_domain}/api/v2/",
                ClientId = _managementClientId,
                ClientSecret = _managementClientSecret
            };

            var tokenResponse = await client.GetTokenAsync(tokenRequest);
            return tokenResponse.AccessToken;
        }
    }

    public class UserProfile
    {
        public int UserId { get; set; }

        public string UserName { get; set; }

        public string Email { get; set; }

        public string PasswordHash { get; set; }
    }
}
```

Could someone please verify that this C# code is correct and whether this is an Auth0 bug or not?

### Environment

- **Version of this library used:** 7.3.0
- **Version of the platform or framework used, if applicable:** .NET Core 3.1 / C# 8 / Visual Studio 2019 (16.7.6)

**Comment by sam-piper** [comment]
I've done more experimentation with the call to the Management API, using manual HttpClient invocation. I've replaced this code in my sample app:

```
// start a new import job using the JSON file
Job job;
await using (var jsonStream = File.OpenRead(jsonFile))
{
    job = await _client.Jobs.ImportUsersAsync(_connectionId, jsonFile, jsonStream, true, jsonFile, false);
}
```

With this code:

```
// start a new import job using the JSON file
using var httpClient = new HttpClient();
httpClient.DefaultRequestHeaders.Authorization = new AuthenticationHeaderValue("Bearer", _accessToken);

var formContent = new MultipartFormDataContent($"---{Guid.NewGuid()}")
{
    { new StringContent(_connectionId), "connection_id" },
    { new StringContent("true"), "upsert" },
    { new StringContent(jsonFile), "external_id" },
    { new StringContent("false"), "send_completion_email" },
    { new ByteArrayContent(Encoding.UTF8.GetBytes(jsonContent)), "users", jsonFile }
};

var response = await httpClient.PostAsync(new Uri($"https://{_domain}/api/v2/jobs/users-imports"), formContent);
response.EnsureSuccessStatusCode();

var responseContent = await response.Content.ReadAsStringAsync();
var job = JsonConvert.DeserializeObject<Job>(responseContent);
```

And it's working perfectly for me now - so either I'm calling the SDK incorrectly, or it looks like there is a bug in the SDK, not the API.

**Comment by frederikprijck** [comment]
Hey @sam-piper ,

Thanks for the follow up.
I will be looking into it now.

**Comment by frederikprijck** [comment]
Hi @sam-piper ,

I tried reproducing this, but I managed to succesfully import the user through our SDK:
![image](https://user-images.githubusercontent.com/2146903/97543483-97734680-19c8-11eb-8843-302dc0f259b9.png)

I used the JSON you provided:
```
[
  {
    "email": "test@user.com",
    "email_verified": true,
    "username": "Test.User",
    "custom_password_hash": {
      "algorithm": "pbkdf2",
      "hash": {
        "value": "$pbkdf2-sha1$i=1000,l=32$d16owXF2G+jw/LpS122jDA$KgqXKR8tdXkTE31rrCsUjVkaDQiPkQQEGd4+u4e3V9g",
        "encoding": "utf8"
      }
    }
  }
]
```
And I used the following Console Application:
```
class Program
{
    static void Main(string[] args)
    {
        MainAsync().Wait();
    }

    static async Task MainAsync()
    {
        var management = new ManagementApiClient("TOKEN", "MYDOMAIN.eu.auth0.com");
        var connectionId = "CONNECTION ID";
        var fileName = $"UserImport-{DateTime.Now:yyyy-MM-dd-HH-mm}.json";
        var filePath = "C:\\path\\to\\users.json";
        var upsert = true;
        var externalId = fileName;
        var sendCompletionEmail = false;

        Job job;
        await using (var fileStream = File.OpenRead(filePath))
        {
            job = await  management.Jobs.ImportUsersAsync(connectionId, fileName, fileStream, upsert, externalId, sendCompletionEmail);
        }

    }
}
```
Could you try using such a simplified example as above ?

**Comment by sam-piper** [comment]
Thanks for the followup. I can't get this minimal example working, the job is created as pending but the user does not get created, same issue as before. I'm not sure what else to try. 

Have you run my original sample console app as provided? Does that work for you as well?

**Comment by frederikprijck** [comment]
I have not, I prefer to keep thinks as minimal as possible. Given you can not get it working with my minimal example means there has to be something else wrong.

Is your connection configured to require usernames?
![image](https://user-images.githubusercontent.com/2146903/97674722-488fe480-1a8e-11eb-929a-14aa80a3ad3c.png)

**Comment by frederikprijck** [comment]
Closing this as no response. Please re open is this continues to be a problem

**Comment by FMagarik** [comment]
Exactly the same issue is happening to me. ImportUsersAsync doesn't seem to be working

**Comment by frederikprijck** [comment]
@FMagarik, does the same issue exists using the code mentioned [here](https://github.com/auth0/auth0.net/issues/443#issuecomment-718473596)?

**Comment by FMagarik** [comment]
> @FMagarik, does the same issue exists using the code mentioned [here](https://github.com/auth0/auth0.net/issues/443#issuecomment-718473596)?

That worked! I saw that another management client constructor was being used so I switched that, but the job is still failing without errors or clues. Also changed the access token for an administrative one, but no luck either. Are there any full code examples for this? I couldn't find any in the documentation. Are there any tools in Auth0's UI to understand why the job is failing with no errors?

**Comment by frederikprijck** [comment]
> Are there any full code examples for this? I couldn't find any in the documentation. 
There isn't anything specific for this, apart from our API documentation. Would you mind adapting the code snippet I shared [here](https://github.com/auth0/auth0.net/issues/443#issuecomment-718473596), to reproduce your issue so I can look into it and see what could be causing it ?

> Are there any tools in Auth0's UI to understand why the job is failing with no errors?
There are always the tenant logs, which can be helpful (select monitor in the navigation, then log).

**Comment by FMagarik** [comment]
> > Are there any full code examples for this? I couldn't find any in the documentation.
> 
> There isn't anything specific for this, apart from our API documentation. Would you mind adapting the code snippet I shared [here](https://github.com/auth0/auth0.net/issues/443#issuecomment-718473596), to reproduce your issue so I can look into it and see what could be causing it ?

If I input my token, domain and connection id into your snippet, then stream a local file with exactly the same content I'm using, it works. Users get created. I also tried uploading that local file into the Users Import/Export extension manually and that also works. I'll keep adding stuff to that snippet, see if I can reproduce the issue from there.

> > Are there any tools in Auth0's UI to understand why the job is failing with no errors?
> 
> There are always the tenant logs, which can be helpful (select monitor in the navigation, then log).

The tenant logs show a successful call to the user import job creation endpoint, returning a 202 with job status pending. Milliseconds later, another log shows the job finished. If I look up that job manually, the status shows as failed. If I look up that job's errors, the endpoint returns 204 No Content. So... No errors. I'm at a loss as to how to diagnose this.

**Comment by FMagarik** [comment]
Ok, I finally got this to work. Instead of creating a file, writing the new user data there and passing its FileStream to ImportUsersAsync, I switched to using a MemoryStream, like so:

```
var fileName = $"userMigrationToAuth0_{DateTime.Now:yyyy-MM-dd-HH-mm}.json";
var content = JsonConvert.SerializeObject(batch);
await using var stream = new MemoryStream(Encoding.UTF8.GetBytes(content));
await mgmtClient.Jobs.ImportUsersAsync("CONNECTION_ID", fileName, stream, true, fileName, false);
```

where batch is a list of custom User objects that get serialized to the JSON structure mentioned in the previous comments, thanks to JsonProperty attributes in their fields.

In the locally-run minimal example that was shared, the FileStream approach works without issues. I suspect the environment in which this is run (in my case a custom Docker container) has an impact on the FileStream object or its capabilities.
